File: src/test/java/io/protobase/f7/formulas/math/ASINTest.java
package io.protobase.f7.formulas.math;

import io.protobase.f7.errors.NAException;
import io.protobase.f7.errors.ValueException;
import io.protobase.f7.models.Grid;
import io.protobase.f7.testutils.TestFormula;
import org.junit.Test;

import static com.google.common.truth.Truth.assertThat;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

public class ASINTest extends TestFormula {
  @Test
  public void testApply() {
    assertThat(ASIN.SELF.apply(null, 10.0)).isEqualTo(Math.asin(10.0));
    assertThat(ASIN.SELF.apply(null, 128731.2)).isEqualTo(Math.asin(128731.2));
    assertThat(ASIN.SELF.apply(null, 11.11)).isEqualTo(Math.asin(11.11));
    assertThat(ASIN.SELF.apply(null, 0.0)).isEqualTo(Math.asin(0.0));
    assertThat(ASIN.SELF.apply(null, 88281.0)).isEqualTo(Math.asin(88281));
    assertThat(ASIN.SELF.apply(null, 2.0)).isEqualTo(Math.asin(2.0));
    assertThat(ASIN.SELF.apply(null, 4.0)).isEqualTo(Math.asin(4.0));
    assertThat(ASIN.SELF.apply(null, -4.0)).isEqualTo(Math.asin(-4.0));
    assertThat(ASIN.SELF.apply(null, -10124.0)).isEqualTo(Math.asin(-10124.0));
  }

  @Test
  public void testApply_stringConversion() {
    assertThat(ASIN.SELF.apply(null, "10.0")).isEqualTo(Math.asin(10.0));
  }

  @Test
  public void testApply_grid() {
    assertThat(ASIN.SELF.apply(null,
        Grid.builder().add(0, 0, -4.0).add(0, 1, "Don't mind me.").build()
    )).isEqualTo(Math.asin(-4.0));
  }

  @Test
  public void test_lookup() {
    ASIN F = new ASIN(lookup, collateralLookup);
    when(collateralLookup.apply(A1, M22_RANGE)).thenReturn(128731.2);
    assertThat(F.apply(A1, M22_RANGE)).isEqualTo(Math.asin(128731.2));
    verify(collateralLookup).apply(A1, M22_RANGE);
    verifyNoMoreInteractions(collateralLookup);
  }

  @Test
  public void testApply_errorsPassThrough() {
    assertThat(ASIN.SELF.apply(null, new ValueException())).isEqualTo(new ValueException());
  }

  @Test
  public void testApply_argumentsMismatch() {
    assertThat(ASIN.SELF.apply(null, "A", "Too many")).isEqualTo(new NAException());
  }
}

File: src/test/js/formulas/math/ASINTest.ts
import { assert } from "chai";
import { it, describe } from "../../testutils/TestUtils";
import { stub } from "sinon";
import { F7ExceptionName } from "../../../../main/js/errors/F7ExceptionName";
import { NAException } from "../../../../main/js/errors/NAException";
import { ValueException } from "../../../../main/js/errors/ValueException";
import { ASIN } from "../../../../main/js/formulas/math/ASIN";
import { Grid } from "../../../../main/js/models/common/Grid";
import { CollateralLookupFunction, LookupFunction } from "../../../../main/js/models/common/Types";
import { Converters } from "../../../../main/js/utils/Converters";
import { CommonModels } from "../../testutils/CommonModels";

describe("ASIN", function () {
  it("should work with valid numbers", function () {
    assert.equal(ASIN.SELF.run(null, 0), Math.asin(0));
    assert.equal(ASIN.SELF.run(null, 0.99), Math.asin(0.99));
    assert.equal(ASIN.SELF.run(null, 1), Math.asin(1));
    assert.equal(ASIN.SELF.run(null, 0.55), Math.asin(0.55));
    assert.equal(ASIN.SELF.run(null, -0.55), Math.asin(-0.55));
  });

  it("should throw NUM error", function () {
    assert.deepEqual(
      Converters.castAsF7Exception(ASIN.SELF.run(null, 1.1)).name,
      F7ExceptionName.NUM
    );
    assert.deepEqual(
      Converters.castAsF7Exception(ASIN.SELF.run(null, -1.1)).name,
      F7ExceptionName.NUM
    );
  });

  it("should work with strings", function () {
    assert.equal(ASIN.SELF.run(null, "1"), Math.asin(1));
  });

  it("should do pass-through errors", function () {
    assert.deepEqual(ASIN.SELF.run(null, new ValueException()), new ValueException());
  });

  it("should use lookup", function () {
    const lookup = stub();
    const collateralLookup = stub();
    const F = new ASIN(lookup as LookupFunction, collateralLookup as CollateralLookupFunction);
    collateralLookup.withArgs(CommonModels.A1, CommonModels.M22_RANGE).returns(0.99);
    assert.deepEqual(F.run(CommonModels.A1, CommonModels.M22_RANGE), Math.asin(0.99));
    assert.isTrue(collateralLookup.calledWith(CommonModels.A1, CommonModels.M22_RANGE));
    assert.equal(collateralLookup.callCount, 1);
    assert.isTrue(lookup.notCalled);
  });

  it("should handle grids", function () {
    const one = Grid.builder().add(0, 0, 0.99).add(0, 1, "A").build();
    assert.deepEqual(ASIN.SELF.run(null, one), Math.asin(0.99));
  });

  it("should return error when argument lengths are wrong", function () {
    assert.deepEqual((ASIN.SELF.run(null) as NAException).name, F7ExceptionName.NA);
    assert.deepEqual((ASIN.SELF.run(null, "A", "B") as NAException).name, F7ExceptionName.NA);
  });
});
